src/logic/safe/api/fetchTokenCurrenciesBalances.ts

```diff
@@ -20,13 +22,85 @@ type FetchTokenCurrenciesBalancesProps = {
 
 export const fetchTokenCurrenciesBalances = async ({
   safeAddress,
-  selectedCurrency,
+  selectedCurrency = 'USD',
   excludeSpamTokens = true,
-  trustedTokens = false,
 }: FetchTokenCurrenciesBalancesProps): Promise<BalanceEndpoint> => {
   const address = checksumAddress(safeAddress)
-  return getBalances(getClientGatewayUrl(), getNetworkId().toString(), address, selectedCurrency, {
-    exclude_spam: excludeSpamTokens,
-    trusted: trustedTokens,
+
+  const items = await toItems(address, selectedCurrency, excludeSpamTokens)
+
+  return {
+    fiatTotal: items.reduce((sum, token) => sum + Number(token.fiatBalance), 0).toString(),
+    items: items.sort((a, b) => {
+      return a.fiatBalance < b.fiatBalance ? 1 : -1
+    }),
+  }
+}
+const WEI_PER = 1_000_000_000_000_000_000
+const exchangeRate = 1
+
+async function toItems(address: string, selectedCurrency: string, excludeSpamTokens: boolean) {
+  const [tokens, prices] = await Promise.all([fetchTokens(address), fetchPrices()])
+
+  const filteredTokens = excludeSpamTokens ? tokens.filter((token) => !/doge/i.test(token.name)) : tokens
+
+  return filteredTokens.map((token) => {
+    const usdConversion = token.symbol === 'cUSD' ? 1 : prices[token.symbol.toUpperCase()] || 0
+    // TODO when we bring back multi fiat support change exchangeRate to the actual rate
+    const fiatConversion = selectedCurrency === 'USD' ? usdConversion : usdConversion * exchangeRate
+    return {
+      tokenInfo: {
+        type: TokenType.ERC20,
+        address: token.contractAddress,
+        decimals: Number(token.decimals),
+        symbol: token.symbol,
+        name: token.name,
+        logoUri: `https://raw.githubusercontent.com/ubeswap/default-token-list/master/assets/asset_${token.symbol}.png`,
+      },
+      balance: token.balance,
+      fiatBalance: new BigNumber(token.balance).dividedBy(WEI_PER).multipliedBy(fiatConversion).toString(),
+      fiatConversion: fiatConversion.toString(),
+    }
   })
 }
+interface BlockscoutTokenBalance {
+  balance: string
+  contractAddress: string
+  decimals: string
+  name: string
+  symbol: string
+  type: string
+}
+
+async function fetchTokens(address: string) {
+  const blockScoutEndpoint = `${getNetworkExplorerInfo().apiUrl}?module=account&action=tokenlist&address=${address}`
+  // retry once if it fails as gonza says that happens sometimes
+  const res = await fetch(blockScoutEndpoint).catch((e) => {
+    console.warn('Fetch error', e)
+    return fetch(blockScoutEndpoint)
```

Comment:
So this would end up being an uncaught exception if it failed again. Is there a sort of final catch we can have for if this also fails?